libbeat/autodiscover/autodiscover.go

```diff
@@ -0,0 +1,243 @@
+package autodiscover
+
+import (
+	"sync"
+
+	"github.com/elastic/beats/libbeat/cfgfile"
+	"github.com/elastic/beats/libbeat/common"
+	"github.com/elastic/beats/libbeat/common/bus"
+	"github.com/elastic/beats/libbeat/logp"
+
+	"github.com/mitchellh/hashstructure"
+)
+
+const debugK = "autodiscover"
+
+// TODO autodiscover providers config reload
+
+// Adapter must be implemented by the beat in order to provide Autodiscover
+type Adapter interface {
+	// TODO Hints
+
+	// CreateConfig generates a valid list of configs from the given event, the received event will have all keys defined by `StartFilter`
+	CreateConfig(bus.Event) ([]*common.Config, error)
+
+	// CheckConfig tests given config to check if it will work or not, returns errors in case it won't work
+	CheckConfig(*common.Config) error
+
+	// RunnerFactory provides runner creation by feeding valid configs
+	cfgfile.RunnerFactory
+
+	// StartFilter returns the bus filter to retrieve runner start triggering events
```

Comment:
So far it seems these are all the same for the current implementation. How will these look for other adapters?

Reply:
I don't know yet :innocent:, I think this code can be used in other parts of beats, not only to detect new modules to lauch, so wanted to keep this open

Reply:
Do we even need the filters methods on the interface? Can the filters be passed when registering the adapter, so to keep the interface somewhat 'smaller'?

Reply:
The adapter is not registered but passed to Autodiscover Manager. If you look at is implementation, it's very small generally, I think it makes sense to keep these filters together with the functions that use it. If in the future we use something different from config we need to update the filters too https://github.com/exekias/beats/blob/411d13a878d64dbe62b6008057bbaa1c207f9e4e/metricbeat/beater/autodiscover.go